Add support for PathSeparator property in Providers

renehernandez opened the pull request:
## PR Summary

Fix #5544.

Add two properties in `ProviderInfo` class: `ItemSeparator` and `AltItemSeparator`.
On windows, the default values for those two properties are `ItemSeparator = '\'` and `AltItemSeparator = '/'`.
On unix, the default values for those two properties are `ItemSeparator = '/'` and `AltItemSeparator = '\'`.

Registry provider is the only exception, both properties for it have the value `\`.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

Commit:
Add PathSeparator property in ProviderInfo

Commit:
Copy PathSeparator property in copy ctor

Commit:
Set PathSeparator for providers

Add PathSeparator values for FileSystem, Function, Variable and Registry
providers
Add corresponding tests

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=8587) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=8587) before we can accept your contribution.<br/><br/>:x: renehernandez [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=8587)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=8587) it.</sub>

Commit:
Fix codefactor issues

Commit:
Fix codefactors issues

Commit:
Run Registry provider test only for windows

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -356,20 +356,25 @@ public string GetHelpMaml(string helpItemName, string path)
             {
-                // %USERPROFILE% - indicate where a user's home directory is located in the file system.
-                string homeDirectory = Environment.GetEnvironmentVariable(Platform.CommonEnvVariableNames.Home);
+                providerInfo.PathSeparator = new[] { "\\", "/" };
```
Perhaps we already have the array in static cache. Below the same.

renehernandez replied:
@iSazonov Good question.  Also, does the FileSystem provider use both separator for *Unix*-like systems as well? Or just the */* character?

I couldn't find anything conclusive on that regard

vexx32 replied:
Unix only uses forward slash, I believe. `\` is a valid filename character, isn't it?

iSazonov replied:
It seems there is Normalization() method in FileSystem provider that allow to use both in PowerShell

renehernandez replied:
@vexx32 I am aware that Unix shells use forward slash only. I guess a better wording for my question would be: Does PowerShell allows the use of both slashes as separators? 😄

vexx32 replied:
@renehernandez had a couple folks in the PowerShell Slack/Discord/IRC channels check it out as I don't have a Linux VM on hand at present, and it seems like the standard PS cmdlets do indeed permit backslashes to be used as a path separator even on Linux.

renehernandez replied:
@iSazonov regarding the initial question, do you suggest to initialize PathSeparator as a reference of  a static (optionally readonly) array instance?

iSazonov replied:
Yes, we need search in our code - perhaps we already have such array.

renehernandez replied:
@iSazonov Regarding the name convention, @mklement0  discussed both on the [issue](https://github.com/PowerShell/PowerShell/issues/5544) and he was leaning towards using `PathSeparator`. Perhaps we could get him to expand on this?

iSazonov replied:
Yes, we need to discuss that will helps PowerShell Committee.

renehernandez replied:
@iSazonov After looking for examples of an array in static cache and not finding anything, I added new static caches in a class next ProviderInfo.

While doing this I realize the need to switch to a `ReadOnlyCollection` instead of a plain array  to avoid situations like the following

```powershell
 $separator = (Get-PSProvider FileSystem).PathSeparator
 $separator[0] = "w"
 (Get-PSProvider FileSystem).PathSeparator | Should -Be @("\", "/") // Fails because it returns @("w", "/")
```

Using `ReadOnlyCollection`, we avoid altogether of the user modifying the PathSeparator property elements

vexx32 replied:
@renehernandez you can define a readonly member something like this, to prevent it from being set by anyone outside the provider class:

```csharp
public string[] PathSeparator { get; private set; } = { "/", "\\" }
```

... _if_ I remember my C# semantics correctly 😉 
You may also want to explicitly declare it as a `readonly` member, but I'm not sure how that would work with where this needs to be set. 😄

renehernandez replied:
@iSazonov @vexx32 There are two different issues to account for here:

1- `PathSeparator` property setter needs to be internally accessible because the different providers are required to set their corresponding value on the property, which they do on the `Start` method receiving a `ProviderInfo` obj as a parameter.

2- The issue that I highlighted in the previous comment will happens if we use `List` or `array` regardless of the scope of the setter property, because the `ProviderInfo` object gets cached per session and thus, we will be assigning a new reference to the same collection every time that `PathSeparator` property is invoked in the same session and modifications to these references will directly modify the original object (which can be seen in the example code that I posted in the above comment). So we either return a read-only object (such as `ReadOnlyCollection` or we return a new object (`List` or `array`) every time the `PathSeparator` property is invoked.

renehernandez replied:
@iSazonov @vexx32 @daxian-dbw Can we resolve this conversation and move forward with the pull request? Or is there anything left that should be changed related to this discussion??

iSazonov replied:
@renehernandez Sorry, I lost the conversation.
We could use:
```c#
public IReadOnlyList<string> PathSeparator { get; } = { "/", "\\" };
```
Perhaps @daxian-dbw has wiser advice.

renehernandez replied:
@iSazonov @vexx32 @daxian-dbw Another alternative would be to use `ImmutableArray` from [System.Collections.Immutable package](https://www.nuget.org/packages/System.Collections.Immutable/). Thinking now, I think it would be a better approach since it gives you a stronger guarantee of safety in multithreaded scenarios. That may not be a problem for the `PathSeparator` property in reality, but it won't hurt.

Any comments?

iSazonov replied:
I think we don't want this if it is an another dll.

renehernandez replied:
@iSazonov the [System.Collections.Immutable](https://www.nuget.org/packages/System.Collections.Immutable/) package is already in use on the powershell code in the [AddType.cs](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs) file, so we wouldn't be introducing a new dependency

iSazonov replied:
@SteveL-MSFT @daxian-dbw What is best type for public API here?

renehernandez replied:
@SteveL-MSFT @daxian-dbw In my perspective deciding on the best API type is currently the only blocker to resolve this conversation and move forward with this pull request. Any input on it?

renehernandez replied:
@SteveL-MSFT @daxian-dbw Any resolution on this topic? I believe this should be relatively simple to tackle

iSazonov commented:
@renehernandez Thanks for your contribution!

I added 'Fix #5544' to link the issue and auto close it after we merge.

The public API should be approved by PowerShell Committee.

Commit:
Change PathSeparator property to ReadOnlyCollection

Fix that changes in a copy of the separator array could affect
subsequent calls of the PathSeparator property

Add test

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-PSProvider.Tests.ps1:
```diff
@@ -29,6 +29,19 @@ Describe "Get-PSProvider" -Tags "CI" {
         }
+
+        It 'changes on copy do not affect original PathSeparator' {
+            $separator = (Get-PSProvider FileSystem).PathSeparator
```
We need tests for all providers. You could use "case" parameter in "It" block.

renehernandez replied:
Done. Also fixed the expected values based on the ordering of default path and alternate path separators used from `StringLiterals` class to build the static arrays

Commit:
Refactor tests to use parameterized data

Use test data accordingly to OS, ie. windows vs Unix-like
Fix expected values for providers for unix-like system

Commit:
Fix codefactors style issues

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and is ok with adding `PathSeparator`

renehernandez commented:
@daxian-dbw Is this going to be merged at all? The changes are pretty small and they shouldn't be to hard to review.

Or is there anything left that should be addressed??

daxian-dbw commented:
@renehernandez Sorry for having this PR slip off my radar. I will do the review later today.

daxian-dbw committed:
Remove extra white space characters

daxian-dbw commented on src/System.Management.Automation/engine/DataStoreAdapterProvider.cs:
```diff
@@ -654,5 +663,35 @@ internal PSNoteProperty GetNotePropertyForProviderCmdlets(string name)
+            {
+                StringLiterals.DefaultPathSeparatorString,
+                StringLiterals.AlternatePathSeparatorString
+            });
```
These 3 static members are in fact duplicates. We should have a single static member instead.
I suggest to add the static separator members to [`Separators`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Utils.cs#L1237), for example:
```
internal static readonly ReadOnlyCollection<string> BackAndForwardSlashSeparators = new ReadOnlyCollection...
internal static readonly ReadOnlyCollection<string> BackSlashOnlySeparators =new ReadOnlyCollection...
```
The same `BackAndForwardSlashSeparators` can be used by `FileSystem`, `Variable`, `Function`, `Env`, `Alias`, `Cert`, and `WSMan`.

renehernandez replied:
Done

daxian-dbw commented on src/System.Management.Automation/namespaces/FunctionProvider.cs:
```diff
@@ -41,6 +41,29 @@ public FunctionProvider()
+        /// <returns>
+        /// The updated ProviderInfo object that holds the provider's configuration.
+        /// </returns>
+        protected override ProviderInfo Start(ProviderInfo providerInfo)
```
Similar method should be added for `Environment`, `Alias`, `Certificate` and `WSMan` providers.

renehernandez replied:
I have yet to add it for the `WSMan` provider. The reason is that `WSMan` provider is in a different namespace, thus would require to change classes in the [Utils file](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Utils.cs) to have public scope

renehernandez replied:
What you should be the outcome here? Should I change the relevant classes' scope?

daxian-dbw replied:
Good point. I think we can use a new instance of `ReadOnlyCollection<string>` for `WSMan`.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-PSProvider.Tests.ps1:
```diff
@@ -24,4 +24,56 @@ Describe "Get-PSProvider" -Tags "CI" {
+
+                $copy[0] = "w"
+
+                (Get-PSProvider $Provider).PathSeparator | Should -Be $Value
```
Please fix the indentations.

renehernandez replied:
Done

daxian-dbw commented on src/System.Management.Automation/engine/DataStoreAdapterProvider.cs:
```diff
@@ -251,6 +251,14 @@ public override string ToString()
+        /// Gets the path separators characters for this provider.
+        /// </summary>
+        /// <returns>
+        /// Array of strings representing the path separators characters for this provider.
```
Comment needs to be updated. It's no longer a string array.

renehernandez replied:
Done

daxian-dbw commented on src/System.Management.Automation/engine/DataStoreAdapterProvider.cs:
```diff
@@ -251,6 +251,14 @@ public override string ToString()
+        /// <returns>
+        /// Array of strings representing the path separators characters for this provider.
+        /// </returns>
+        public ReadOnlyCollection<string> PathSeparator { get; internal set; } = new ReadOnlyCollection<string>(new List<string>());
```
Just wondering, should we allow string that is more than 1 chars to be used as the path separator? Maybe `ReadOnlyCollection<char>`?

daxian-dbw replied:
Also, as for the default value of this property, maybe it should be `null` by default?

renehernandez replied:
Do you want me to apply these changes?

iSazonov replied:
I agree with the changes.

daxian-dbw replied:
I want to be discreet here as it's a public API.
@SteveL-MSFT @joeyaiello what do you think?

SteveL-MSFT replied:
@daxian-dbw's feedback makes sense to me.  I'm not aware of any situation where the separator is more than a single character.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-PSProvider.Tests.ps1:
```diff
@@ -24,4 +24,56 @@ Describe "Get-PSProvider" -Tags "CI" {
+                        }
+                        else {
+                            $nonWindowsTestCases
+                        }
```
Not blocking. It's better to change to the following, so we can avoid unnecessary code on a specific platform.
```
$testCases = if ($IsWindows) {
    @{Provider = 'FileSystem'; Value = @("\", "/")}
    ...
} else {
    @{Provider = 'FileSystem'; Value = @("/", "\")}
    ...
}
```

renehernandez replied:
How is that different to what I am currently doing?

iSazonov replied:
Dongbo mean that currently we create extra variable $windowsTestCases or  $nonWindowsTestCases  - we need one not both.

renehernandez replied:
Done

daxian-dbw replied:
In your original code, both `$windowsTestCases` and `$nonWindowsTestCases` will be created and initialized no matter what platform the test is running on.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-PSProvider.Tests.ps1:
```diff
@@ -24,4 +24,56 @@ Describe "Get-PSProvider" -Tags "CI" {
+        }
+
+        It 'copying and modifying values does not affect PathSeparator property in <Provider> provider' -TestCases $testCases {
+            param ($Provider, $Value)
```
This test case doesn't seem to be very interesting. Changing the copy of a collection doesn't affect the original collection.

renehernandez replied:
Now that we're using a `ReadOnlyCollection` probably not, I put it in place to provide a case that if we were to use an array instead, it would be possible to modify the content of the PathSeparator in the same powershell session

daxian-dbw replied:
Fair point.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-PSProvider.Tests.ps1:
```diff
@@ -24,4 +24,56 @@ Describe "Get-PSProvider" -Tags "CI" {
+
+            $separator = (Get-PSProvider $Provider).PathSeparator
+
+            { $separator[0] = "w" } | Should -Throw
```
Maybe add one test to check that the public `PathSeparator` is get-only

renehernandez replied:
Done

daxian-dbw commented:
@renehernandez When you are addressing my comments, can you please add the `[Feature]` tag to the last commit's message? I would like the CI to run all feature tests for this PR.

Commit:
Apply changes requested by daxian

Add PathSeparator property to Env, Cert and Alias provider
Add test for property being read-only
Consolidate separator list in Separators class

Commit:
Remove intermediate variables

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-PSProvider.Tests.ps1:
```diff
@@ -24,4 +24,64 @@ Describe "Get-PSProvider" -Tags "CI" {
+                                @{Provider = 'Function'; Value = @("/", "\")}
+                                @{Provider = 'Alias'; Value = @("/", "\")}
+                                @{Provider = 'Environment'; Value = @("/", "\")}
+                                @{Provider = 'Certificate'; Value = @("/", "\")}
```
The `Certificate` and `WSMan` providers are windows-only.

renehernandez replied:
I have removed them from the nonWindows case. I realized at that point that the `WSMan` provider is not enabled when these tests are run. Should I add the corresponding tests for `WSMan` at [ConfigProvider.Tests.ps1](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.WSMan.Management/ConfigProvider.Tests.ps1)?

daxian-dbw replied:
If `WSMan` provider is changed in this PR, then it should also be tested along with the rest.

renehernandez replied:
Added tests for `ItemSeparator`/`AltItemSeparator` at https://github.com/PowerShell/PowerShell/pull/8587/files#diff-c2f140f155813763bf7a3492a6a03196

daxian-dbw reviewed (commented):
Mostly looks good to me. Thanks! Let's wait for Steve and Joe's feedback on `string` vs. `char`.
Can you please rebase when push your next commit?

> can you please add the `[Feature]` tag to the last commit's message? I would like the CI to run all feature tests for this PR.

And please, add the `[Feature]` tag to the last commits's message. Thanks

Commit:
Add WSMan path separator property

Remove invalid test cases for nonwindows os

iSazonov commented on src/System.Management.Automation/engine/DataStoreAdapterProvider.cs:
```diff
@@ -257,7 +257,7 @@ public override string ToString()
         /// Read-only collection of strings representing the path separators characters for this provider.
         /// </returns>
-        public ReadOnlyCollection<string> PathSeparator { get; internal set; } = new ReadOnlyCollection<string>(new List<string>());
+        public ReadOnlyCollection<string> PathSeparator { get; set; } = new ReadOnlyCollection<string>(new List<string>());
```
Why we need public set?

renehernandez replied:
Because the `WSMan` provider is in different namespace, thus the `set` property cannot be internal in order to be used there

daxian-dbw replied:
Ah, I forgot the S.M.A.dll is not internal visible to `Microsoft.WSMan.Management`.
We need to figure out a different way to do this then, because this property shouldn't have a public setter.

daxian-dbw commented:
@renehernandez I think we need to implement the `PathSeparator` property in a different way.
Here is my suggestion:
Add a `public virtual` get-only property `PathSeparators` in `CmdletProvider`. The default implementation can be this:
```
public virtual ReadOnlyCollection<string> PathSeparators => Utils.EmptyReadOnlyCollection<string>();
```
Then, add a public-getter-private-setter property `PathSeparators` in `ProviderInfo`. The default implementation can be this:
```
public ReadOnlyCollection<string> PathSeparators { get; private set; }
```
Then, in the `ProviderInfo.CreateInstance`, after casting `providerInstance` to `CmdletProvider`, we set the `PathSeparators` property in the Provider instance:
```
Provider.CmdletProvider result = providerInstance as Provider.CmdletProvider;
PathSeparators = result.PathSeparators;
```
Then, in each provider implementation, override that property to provide the supported separators.

renehernandez commented:
@daxian-dbw Are you suggesting a rename as part of the change? Should the property be plural, i.e. `PathSeparators`?

daxian-dbw commented:
```
PS:28> [System.IO.Path] | gm -Static
...
AltDirectorySeparatorChar   Property   static char AltDirectorySeparatorChar {get;}
DirectorySeparatorChar      Property   static char DirectorySeparatorChar {get;}
InvalidPathChars            Property   static char[] InvalidPathChars {get;}
PathSeparator               Property   static char PathSeparator {get;}
VolumeSeparatorChar         Property   static char VolumeSeparatorChar {get;}
```

@renehernandez If we look at how `System.IO.Path` name those properties, I think plural would be better as we return a collection. But let me check with the committee before we move forward.

daxian-dbw commented:
@PowerShell/powershell-committee Need your another review on two items:
1. Since the new property `PathSeparator` returns a collection (ReadOnlyCollection), shall we rename it to `PathSeparators` (plural)?
2. Do we allow a string of more than 2 characters used as a path separator in PowerShell providers? If not, maybe make the property return `ReadOnlyCollection<char>`?

mklement0 commented:
Here's my 2¢:

* Re 1: I know it's late in the game, but, seeing the CoreFx approach of providing `DirectorySeparatorChar` and `AltDirectorySeparatorChar` separately, I wonder if we should similarly provide two _scalar_ properties, `PathSeparator` and `AltPathSeparator` - given that my guess is that users will typically want _the_ official separator, even though there _may_ be an alternative (which, as far as native data stores go, really only applies to the filesystem on Windows).

  * As an aside: `\` currently _de facto_ works on Unix too, but shouldn't, because `\` is a legal filename character -  see #3666

* Re 2: My vote is for `string` rather than `char`, for the following reasons:
  * Given that the property will (also) be consumed from PowerShell code, there is no advantage to using `char` over `string`, given that `char` is not a first-class type in PowerShell code, where even single characters are by default represented as `string`s.
  * It leaves the door open for providers with multi-character separators - though that is admittedly not very likely to happen.

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this, based on the current [code](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Utils.cs#L1226), PowerShell expects separators to be chars so supporting it as string would add more changes.  We are ok with calling it `PathSeparators` to denote being a collection.  We are also ok with adding the new virtual member with default implementation as the impact is expected to be minimal, but fix is clear.

mklement0 commented:
Thanks, @SteveL-MSFT, just to clarify re _naming_: 

The previously agreed-upon name was `PathSeparator`, so that means the name pair should be `PathSeparator` and `AltPathSeparator`, correct?

The names `[Alt]DirectorySeparatorChar` wouldn't be provider-neutral; there are no _directories_ in the registry, for instance. 

By contrast, what the property returns is the character used to separate the items forming a _path_  (which is a provider-neutral concept, the potential for confusion with the poorly named `[IO.Path.PathSeparator]` notwithstanding - see the original discussion in #5544)

mklement0 commented:
@SteveL-MSFT: Or did I misread the following?

> We suggest with just DirectorySeparator and AltDirectorySeparator (default should be same as IO.Path.DirectorySeparatorChar and IO.Path.AltDirectorySeparatorChar to match OS).

Please clarify if there should be only a single, array-valued property or whether there should be two scalar properties.

SteveL-MSFT commented:
@mklement0 Two scalar properties that replace the need for a single collection.  The affinity to calling it DirectorySeparator (sans Char) is that it is more similar to DirectorySeparatorChar than PathSeparator in how they function.  We briefly considered other names, but nothing was a clear winner.

mklement0 commented:
Thanks for the clarification, @SteveL-MSFT.

Allow me another thought regarding the names chosen:

* The very motivation for creating the property/ies at hand was to provide a _provider-neutral_ way to determine the character that is used to separate any given provider's item names when forming a path.

* Having the name `Directory` in the property name directly contradicts that goal and is likely to cause confusion.

* If avoiding confusing with the poorly named `[System.IO.Path]::PathSeparator` is a must (I personally don't think it is - it's CoreFx's mistake, and many PowerShell users may never even learn of its existence) the provider-neutral abstraction from `[Alt]DirectorySeparator` would be **`[Alt]ItemSeparator`**.

SteveL-MSFT commented:
@mklement0 We appreciate your feedback.  @PowerShell/powershell-committee discussed this further and given that the `*-Item` cmdlets work against PSProviders, calling it `ItemSeparator` and `AltItemSeparator` makes sense and ok with adding those two scalars.

daxian-dbw commented:
Given the approval from the committee, the implementation recommended from https://github.com/PowerShell/PowerShell/pull/8587#issuecomment-471345280 needs to be updated:

Add two `public virtual` get-only properties in `CmdletProvider` with the default implementation:
```
public virtual char ItemSeparator => Path.DirectorySeparatorChar;
public virtual char AltItemSeparator => Path.AltDirectorySeparatorChar;
```
Then, add two public-getter-private-setter properties in `ProviderInfo`:
```
public char ItemSeparator { get; private set; }
public char AltItemSeparator { get; private set; }
```
Then, in the `ProviderInfo.CreateInstance`, after casting `providerInstance` to `CmdletProvider`, we set the above two properties in the Provider instance:
```
Provider.CmdletProvider result = providerInstance as Provider.CmdletProvider;
ItemSeparator = result.ItemSeparator;
AltItemSeparator = result.AltItemSeparator;
```
Then, in each provider implementation, override that two properties to provide the supported separators.

iSazonov commented:
@renehernandez Could you implement the proposal? If you need a help fill free to ask.

renehernandez commented:
@daxian-dbw The only question with the changes is: What should be the value for `AltItemSeparator` be for the `RegistryProvider`? The empty char?

iSazonov commented:
I'd expect that AltItemSeparator is the same ItemSeparator.

mklement0 commented:
Agreed, @iSazonov. It is what CoreFx's `[IO.Path]::AltDirectorySeparatorChar` does on Unix (it reports `/` too); from the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.altdirectoryseparatorchar?view=netcore-2.2):

> This field can have the same value as `DirectorySeparatorChar`.

Commit:
[Feature] Rewrite separator properties

Remove unused overrides for Start method
Add tests for WSMan in ConfigProvider tests

Commit:
[Feature] Apply codefactor fixes

Remove unused properties added to internal Separator class
Fix xml documentation

Commit:
[Feature] More codefactor fixes

Remove unused namespace
Fix xml documentation tag
Remove base prefix from expression body

renehernandez commented:
@daxian-dbw All the 16 codefactor issues are categorized as **style** and they were not introduced by my changes. What is the action there?

daxian-dbw commented:
@renehernandez Thanks for getting back and finishing up this PR! No action needed. The `Hungarian notation` rule is not useful in most cases.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -354,21 +355,24 @@ public string GetHelpMaml(string helpItemName, string path)
         {
             // Set the home folder for the user
-            if (providerInfo != null && string.IsNullOrEmpty(providerInfo.Home))
+            if (providerInfo != null)
```
I think you can revert changes in this file.

renehernandez replied:
Done

daxian-dbw commented on src/System.Management.Automation/namespaces/ProviderBase.cs:
```diff
@@ -1384,6 +1385,16 @@ public PSHost Host
+        /// <summary>
+        /// Gets the alternate item separator character for this provider.
+        /// </summary>
+        public virtual char AltItemSeparator => Path.AltDirectorySeparatorChar;
```
The file system provider (and others) doesn't override these two properties, and that means on Unix platforms, both `ItemSeparator` and `AltItemSeparator` for the file system provider (and others) are forward slash `/`.
Theoretically, this probably is correct. But it's not what those providers behave today -- file system provider, env provider, function provider, alias provider, variable provider, they all accept and normalize `/` and `\` on Unix plats.

Yes, today the registry provider also accepts and normalizes `/` and `\`. But we agreed that's not right and should be corrected in future. Therefore, the override in registry provider makes senses.
If we agree to make both `ItemSeparator` and `AltItemSeparator` the forward slash `/` for other providers on Unix plats, it would imply that we are going to change those providers to only accept `/` on Unix, which I don't think is true ...

daxian-dbw replied:
/cc @SteveL-MSFT @JamesWTruher @mklement0 @iSazonov for input.

vexx32 replied:
Given that `\` is an acceptable file or folder name character on Unix filesystems, unless I'm mistaken, shouldn't that be fixed?

mklement0 replied:
I agree that these properties should always reflect the de-facto behavior of the providers.

Separately, I personally think that the filesystem provider _shouldn't_ accept `\` on Unix-like platforms - but that's probably a topic for #6745.
Consistent cross-platform behavior is a wonderful thing in principle, but not if it comes at the expense of making native features inaccessible / cumbersome to access. I wish we had gone with just `/` on Unix and told people to consistently use `/` in cross-platform code.

daxian-dbw replied:
> Given that \ is an acceptable file or folder name character on Unix filesystems, unless I'm mistaken, shouldn't that be fixed?

Well, make the file system provider only accept `/` on Unix plats would definitely a breaking change that will affect many existing scripts.
So maybe the fix should be: if the path contains `\` in it, then first try resolving the path literally (not treating `\` as a directory separator), and if nothing is found, then treat `\` as a directory separator. If we go with a fix like this, then `\` is still accepted as a item separator in file system provider, under some conditions though.

iSazonov replied:
Having / on Unix and \ on Windows is obvious that a compromise is needed. And it is already implemented with some side effects. 
It is also clear that we will not be able to get rid of all the side effects. 
It is also clear that users will use their usual value _in console_ and most likely it will correspond to their main platform (but not the platform on which they are now located).
For scripts, we need to describe in docs how to create portable paths and implement something like Test-Path -Portable and perhaps [portablepath].

I know only one way to create a portable path - Join-Path.
__Based on this and the above, I would generally forbid the path separators, so as not to provoke users to create non-portable scripts.__

SteveL-MSFT replied:
I agree with @iSazonov, I use `Join-Path` now whenever I have a file system path as it's the predictable way to have a correct path for the target system.  Perhaps as part of the FileSystemProvider v2 thinking, we should only support `/` on non-Windows platforms, but that's a separate [discussion](https://github.com/PowerShell/PowerShell/issues/6833).  For the registry provider there's already an [issue](https://github.com/PowerShell/PowerShell/issues/5536) where it can't read a path that contains a `/` which is valid for the registry.  For this PR, it might be better to default to `Path.DirectorySeparatorChar` for `AltItemSeparator` and require explicit setting to `Path.AltDirectorySeparatorChar`.

mklement0 replied:
@SteveL-MSFT: While `Join-Path` is a clean and robust solution, it is also verbose and slow. 

**It's easy to create portable paths directly: use `/` on all platforms**, which even .NET and the Windows API understand on Windows (support is inconsistent in `cmd`).

Additionally, on both platforms it is _benign_ to accidentally double `/` when building paths with string concatenation (e.g., `Get-Item /tmp//foo` works just fine).

I think preventing the direct use of path separators would be ill-advised.

I agree that it's a separate discussion, but note that, just like with the registry provider, the current filesystem provider cannot access files with names such as `a\b` on Unix, where such filenames are legal.

vexx32 replied:
Is there a `Resolve-Path -Normalize` to convert slashes? That'd be really nice... 🤔

mklement0 replied:
@vexx32 , it's what both `Resolve-Path` (resolving in terms of PS drives) and `Convert-Path`  (resolving in terms of native paths) already do, implicitly.

While `Convert-Path` invariably resolves to a _full_ (and native) path, `Resolve-Path -Relative` works with relative paths too.

The limitation is that both cmdlets currently require the target path to _exist_ - see #5730.

 Related: #6197, #8435 , #3501

vexx32 replied:
Yep I've been trying to remove that requirement, but I can't for the life of me find the code that does that in certain cases. :(

renehernandez replied:
So, what is the action to be applied from this conversation?

SteveL-MSFT replied:
Re-tagging for @PowerShell/powershell-committee review

Commit:
[Feature] Remove no longer applicable changes

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this, we should keep use of ItemSeparator = DirectorySeparator and AltItemSeparator = AltDirectorySeparator by default.  Registry Provider has a bug and should be fixed (as seperate PR).  Filesystem accepting both / and \ on all systems is an explicit cross platform experience decision for PowerShell.

mklement0 commented:
@SteveL-MSFT: Please see #9244

daxian-dbw commented on src/System.Management.Automation/namespaces/ProviderBase.cs:
```diff
@@ -1384,6 +1385,16 @@ public PSHost Host
+        /// <summary>
+        /// Gets the alternate item separator character for this provider.
+        /// </summary>
+        public virtual char AltItemSeparator => Path.AltDirectorySeparatorChar;
```
OK, given the committee's decision, we need to make the following change to this property.
@renehernandez Could you please update the code the test? Thanks!

```
        /// <summary>
        /// Gets the alternate item separator character for this provider.
        /// </summary>
        public virtual char AltItemSeparator =>
#if UNIX
            Utils.Separators.Backslash[0];
#else
            Path.AltDirectorySeparatorChar;
#endif
```

renehernandez replied:
@daxian-dbw Will do

renehernandez replied:
@daxian-dbw Any updates regarding the latest changes?

Commit:
[Feature] Switch to use Utils.Separators in Unix environment

Rebase from master

daxian-dbw reviewed (approved):
Just realized new commits were pushed. This looks good to me. Thank you @renehernandez!

iSazonov commented on src/System.Management.Automation/namespaces/ProviderBase.cs:
```diff
@@ -1384,6 +1385,21 @@ public PSHost Host
+        /// </summary>
+        public virtual char AltItemSeparator =>
+#if UNIX
+            Utils.Separators.Backslash[0];
```
@daxian-dbw Why don't we use `Path.DirectorySeparatorChar`?
https://github.com/dotnet/corefx/blob/78c4c8bea5121408b0b420c2aa029cdd992a0b26/src/Common/src/CoreLib/System/IO/PathInternal.Unix.cs#L14

iSazonov commented:
> today the registry provider also accepts and normalizes / and \. But we agreed that's not right and should be corrected in future. 

Have we a tracking issue?

daxian-dbw commented:
The issue tracking registry provider is https://github.com/PowerShell/PowerShell/issues/5536

daxian-dbw commented:
@renehernandez Thank you again for your contribution and patience!
